runtime/src/lib.rs

```diff
@@ -752,6 +1003,115 @@ pub type Executive = frame_executive::Executive<
 	AllPalletsWithSystem,
 >;
 
+#[cfg(feature = "std")]
+fn get_utf8_string(string: &str) -> Result<&str, &str> {
+	Ok(string)
+}
+#[cfg(not(feature = "std"))]
+fn get_utf8_string(string: &Vec<u8>) -> Result<&str, &str> {
+	str::from_utf8(&string[..]).map_err(|_| "Lo siento")
+}
+
+
+fn is_valid(category: &Categories, data: &Vec<u8>, proto_references: &Vec<Hash256>) -> bool {
+	match category {
+		Categories::Text(sub_categories) => match sub_categories {
+			TextCategories::Plain => str::from_utf8(data).is_ok(),
+			// REVIEW - does a Json have to be a `serde_json::Map` or can it `serde_json::Value`?
+			TextCategories::Json => serde_json::from_slice::<serde_json::Map<String, serde_json::Value>>(&data[..]).is_ok()
+		},
+		Categories::Trait(trait_hash) => match trait_hash { // Non Capisco Cosa Fare Qui!!!
+			Some(_) => false,
+			None => {
+				let Ok(trait_struct) = Trait::decode(&mut &data[..]) else { // REVIEW - is `&mut *data` safe?
+					return false;
+				};
+
+				if trait_struct.name.len() == 0 {
+					return false;
+				}
+
+				trait_struct.records.windows(2).all(|window| {
+					let (record_1, record_2) = (&window[0], &window[1]);
+					let (Ok(a1), Ok(a2)) = (get_utf8_string(&record_1.0), get_utf8_string(&record_2.0)) else { // `a1` is short for `attribute_1`, `a2` is short for `attribute_2`
+						return false;
+					};
+
+					let (Some(first_char_a1), Some(first_char_a2)) = (a1.chars().next(), a2.chars().next()) else {
+						return false;
+					};
+
+					(first_char_a1.is_alphabetic() && first_char_a2.is_alphabetic()) // ensure first character is an alphabet
+						&&
+						(a1 == a1.to_lowercase() && a2 == a2.to_lowercase()) // ensure lowercase
+						&&
+						a1 <= a2 // ensure sorted. Note: "Strings are ordered lexicographically by their byte values ... This is not necessarily the same as “alphabetical” order, which varies by language and locale". Source: https://doc.rust-lang.org/std/primitive.str.html#impl-Ord-for-str
+						&&
+						a1 != a2 // ensure no duplicates
+				})
+			},
+		},
+		Categories::Shards(shards_script_info_struct) => {
+			let format = shards_script_info_struct.format;
+			let requiring = &shards_script_info_struct.requiring;
+			let implementing = &shards_script_info_struct.implementing;
+
+			let all_required_trait_impls_found = requiring.iter().all(|shards_trait| {
+				proto_references.iter().any(|proto| {
+					if let Some(trait_impls) = pallet_protos::TraitImplsByShard::<Runtime>::get(proto) {
+						trait_impls.contains(shards_trait)
+					} else {
+						false
+					}
+				})
+			});
+
+			let all_traits_implemented_in_this_shards = implementing.iter().all(|shards_trait| {
+				match format {
+					ShardsFormat::Edn => {
+						// TODO - How do I check if this Shards is implementing this Trait - Giovanni Petrantoni
```

Comment:
@sinkingsugar - how exactly do I verify if this shards script is implementing all the traits it specified in `Categories::Shards::implementing` 😅 ? Thanks :)

Reply:
you don't for now it's fine